Product Collection: Introduce "Related Products" Collection

ObliviousHarmony opened the pull request:
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

This pull request introduces a new "Related Products" collection. The new collection allows merchants to select a product and the collection will list products related to the selected one. When a product context is available, such as on template pages, and in the single product block, it will use that as a fallback.

In pursuit of this, I have also added a new mechanism for implementing collection types on the server. 

Closes #49335

<!-- Begin testing instructions -->

### How to test the changes in this Pull Request:

<!-- Include detailed instructions on how these changes can be tested. Review and follow the guide for how to write high-quality testing instructions. -->

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

#### Explicit Reference

1. Add a new Product Collection to a post and select "Related Products" as the type.
2. Confirm that you are prompted for a product and select one.
3. You should be presented the same content as the "Related Products" block in both the frontend and backend.

#### Single Product Block

> https://github.com/woocommerce/woocommerce/issues/51352 documents a bug with this flow unrelated to the PR.

1. Add a new "Single Product" block and select the same product you did in the previous set of tests.
2. Add a Product Collection somewhere _inside_ the block and select the "Related Products" collection.
3. Confirm that you are _not_ prompted to enter a product.
4. You should be presented the same content as the "Related Products" block in both the frontend and backend.

#### Product Template

1. Open the "Single Product" template in the site editor.
2. Add a Product Collection above or below the "Related Products" block that is already present in the template and selected the "Related Products" collection.
3. Confirm that you are _not_ prompted to enter a product.
4. Confirm that when the Product Collection is selected the "Preview" icon is visible.
5. Confirm that a random set of products are returned.
6. Visit a product page that renders the template. Confirm that the blocks both show related products.

<!-- End testing instructions -->

### Changelog entry

<!-- You can optionally choose to enter a changelog entry by checking the box and supplying data. -->

-   [x] Automatically create a changelog entry from the details below.

<!-- If no changelog entry is required for this PR, you can specify that below and provide a comment explaining why. This cannot be used if you selected the option to automatically create a changelog entry above. -->

-   [ ] This Pull Request does not require a changelog entry. (Comment required below)

<details>

<summary>Changelog Entry Details</summary>

#### Significance

<!-- Choose only one -->

-   [ ] Patch
-   [x] Minor
-   [ ] Major

#### Type

<!-- Choose only one -->

-   [ ] Fix - Fixes an existing bug
-   [x] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message <!-- Add a changelog message here -->

New "Related Products" Product Collection type.

</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment <!-- If your Pull Request doesn't require a changelog entry, a comment explaining why is required instead -->

</details>

ObliviousHarmony committed:
Fixed PHP 7.4 Compatibility

The splat operator does not support associative arrays until PHP 8.1.

ObliviousHarmony committed:
Added Related Products Collection Boilerplate

ObliviousHarmony committed:
Added `woocommerceRelatedTo` Collection Parameter

This parameter allows the block to query for products that are related
to the given product IDs.

ObliviousHarmony committed:
Added Collection Related Product Filter

Using `woocommerceRelatedTo`, queries may restrict the products returned
to those that are related to the given product ID(s).

ObliviousHarmony committed:
Removed Unnecessary Tracking

Commit:
Add changefile(s) from automation for the following project(s): woocommerce-blocks, woocommerce

ObliviousHarmony committed:
Merge branch 'trunk' into add/49335-related-products-collection

ObliviousHarmony committed:
Removing Misunderstood Option

ObliviousHarmony commented on plugins/woocommerce-blocks/assets/js/blocks/product-collection/collections/related-to.tsx:
```diff
@@ -0,0 +1,63 @@
+	description: __( 'Recommend products like this one.', 'woocommerce' ),
+	keywords: [ 'related', 'product collection' ],
+	scope: [],
+	preview: {
```
This will get removed when I add the reference!

imanish003 replied:
Sounds good! JFYI `usesReference` was introduced in [#49796](https://github.com/woocommerce/woocommerce/pull/49796). Therefore, It was already available when you were working on this feature. For "Related Products" collection, we can set `usesReference` to `["product"]`. If you're curious to see how it works, there's an example [here](https://github.com/woocommerce/woocommerce/blob/71ed41c2736ea3fefea00609c4bc66d05a5ae06e/docs/product-collection-block/register-product-collection.md#L249-L277).

ObliviousHarmony replied:
Well, you're not wrong :smile:

ObliviousHarmony commented on plugins/woocommerce/src/Blocks/BlockTypes/ProductCollection.php:
```diff
@@ -747,24 +754,35 @@ private function get_final_frontend_query( $query, $page = 1, $is_exclude_applie
+		);
+
+		$handpicked_products = $query['handpicked_products'] ?? array();
+		$related_products    = $this->get_related_products( $query['related_to'], $common_query_values['posts_per_page'] );
```
One concern I have about using `'posts_per_page'` is that additional filters would receive only `n` products and filter to less, even if the related products would be more.

Should we perhaps set this to something higher (or unlimited?) and then let the other filtering take place? This could work, however, some sites have an unreasonably large number of products and that might not be great in terms of performance.

imanish003 replied:
Hmm Good question! It's valid concern about the potential limitations of using 'posts_per_page' in this context. Thinking out loud:
- As you suggested, We can try setting it to a bit higher value (e.g. 200) and then let the other filtering take place. This way, we won't have to worry about performance issues.
	- But it's really hard to find a value that would work in all cases. Ideal solution would be to have `wc_get_related_products` to return all related products. By fetching all related products, we can ensure that subsequent filters (like price range or attributes) have the full set of related products to work with.
- Maybe we can try finding alternative solution which doesn't include using `wc_get_related_products` to get all related products?
- Once we call `wc_get_related_products`, Maybe we can use filters which allow us to modify the query itself to make it more performant?
	- AFAIU `wc_get_related_products` internally end up calling [get_related_products](https://github.com/woocommerce/woocommerce/blob/58d5c95dfeb2e73fd96dd36044fc7ff172bb4c30/plugins/woocommerce/includes/data-stores/class-wc-product-data-store-cpt.php#L1460) function. This function have `woocommerce_product_related_posts_query` filter. Maybe we can use this filter to modify the query to make it more performant?
- Maybe we can check if it's possible to use [posts_clauses](https://github.com/woocommerce/woocommerce/blob/58d5c95dfeb2e73fd96dd36044fc7ff172bb4c30/plugins/woocommerce/src/Blocks/BlockTypes/ProductCollection.php#L121) to update the final WP_Query, similar to how we are doing for Price Range filter as you can see in [add_price_range_filter_posts_clauses](https://github.com/woocommerce/woocommerce/blob/58d5c95dfeb2e73fd96dd36044fc7ff172bb4c30/plugins/woocommerce/src/Blocks/BlockTypes/ProductCollection.php#L1570) function. 

It feels like a tough problem to solve, but we need to find a solution to make it work for all sites. Maybe we can start with easier approach and then move towards more complex approach. Let me know your thoughts 🙇‍♂️ 

CC: @kmanijak In case you have other suggestions 😅

ObliviousHarmony replied:
> Maybe we can try finding alternative solution which doesn't include using `wc_get_related_products` to get all related products?

The benefit of using `wc_get_related_products()` is that it's the standard way to get other products related to a product. It has lots of filters already in place and I'm sure there's plenty of extensions that modify the behavior to suit their individual project. Ideally any solution we come up with should use this same functionality so that there's parity with existing "Related Products" behavior.

> Once we call `wc_get_related_products`, Maybe we can use filters which allow us to modify the query itself to make it more performant?

I'm not _sure_ what kind of performance characteristics related products have. I just know it's a second query that runs to get a list of products to include.

> It feels like a tough problem to solve, but we need to find a solution to make it work for all sites. Maybe we can start with easier approach and then move towards more complex approach. Let me know your thoughts 🙇‍♂️

I'd like for us to approach this more conservatively at first. We should keep using `wc_get_related_products()` and make any improvements that we need to later on when they become apparent. I don't really know what kind of performance implications this might have and it's probably something we should test on a large store. The main concern I have here is that it's a second query (get related product IDs and use them in the main query) but I'd wager the two query approach is faster than something with a bunch of convoluted joins to find IDs that exist in a subset etc etc etc

For what it's worth, there _is_ a cache in place for related products already so all of this is probably fine :smile: To the original point, let's see what removing the limit entirely does and how that impacts the overall performance of the collection.

imanish003 replied:
> Ideally any solution we come up with should use this same functionality so that there's parity with existing "Related Products" behavior.

Aha! Good point. Makes sense to me. 

> I'm not sure what kind of performance characteristics related products have. I just know it's a second query that runs to get a list of products to include.

Maybe I am overthinking it 🤷🏻 
- I thought calling `wc_get_related_products` with large limit will have performance implications. 
- Also, Having lot of ID's in `post__in` will have performance implications, as we know that `post__in` is processed using IN clause in SQL. 

But as you said, we can probably improve it later if needed.

> To the original point, let's see what removing the limit entirely does and how that impacts the overall performance of the collection.

Sounds good to me. We can always improve it later if needed 😀

xristos3490 replied:
> We should keep using wc_get_related_products() and make any improvements that we need to later on when they become apparent.

I'm also in favor of this.

> For what it's worth, there is a cache in place for related products already so all of this is probably fine

I believe it's a good idea to set a limit here, even if it affects the query builder. 

Looking at the code in `wc_get_related_products()`, it seems that the transients [are written](https://github.com/woocommerce/woocommerce/blob/d2fb60a5d12705cc984a79151c1a284b2af96a52/plugins/woocommerce/includes/wc-product-functions.php#L999) _regardless_ of the limit, and each limit request generates a new one if there are "not enough cached". 

As of now, every store is flooded with transients that contain _five_ ids, yes. If we opt for a limitless query, considering a large catalog with thousands of IDs in each transient, it could potentially affect resources(?) It's definitely a tough problem. I suggest using a reasonable limit, not 5, but not unlimited. Perhaps 100 would be a good compromise?

ObliviousHarmony replied:
I've gone with 100 in my latest push. We can always change the number later and I think the point @xristos3490 is a _really_ good one. There are plenty of sites with way more products than you can imagine and that would be wild to potentially cache every permutation of relationship between products.

ObliviousHarmony reviewed (commented):
Before this PR can be marked as ready we will need to integrate the product context. Once https://github.com/woocommerce/woocommerce/pull/50164 has been merged we can do that.

In consideration of that, I'm not sure how best to approach the inclusion of the product reference. These are the options I have considered:

- Serialize `woocommerceRelatedTo` with the query from the client.
- Allow `true` in `woocommerceRelatedTo` to indicate that it should rely on context and dynamically select that on the server.
  - Alternatively, a string like `context` might be more descriptive?

I am inclined towards doing the second option but some feedback from a reviewer would be nice!

ObliviousHarmony committed:
Linting Fix

ObliviousHarmony committed:
Fixed Test

imanish003 committed:
Merge branch 'trunk' of https://github.com/woocommerce/woocommerce into add/49335-related-products-collection

imanish003 commented:
Hey @ObliviousHarmony,

Great work! 🚀 Thanks for creating a draft PR for early feedback. I am sure that it will lead to a great implementation of "Related Products" collection. 🙌🏻

Although the current implementation works well, I wanted to share my thoughts on how the "Related Products" collection could work ideally.

> In consideration of that, I'm not sure how best to approach the inclusion of the product reference. These are the options I have considered:
> 
> - Serialize `woocommerceRelatedTo` with the query from the client.
> - Allow `true` in `woocommerceRelatedTo` to indicate that it should rely on context and dynamically select that on the server.
>     - Alternatively, a string like `context` might be more descriptive?

Everything starts with specifying `usesReference` to `["product"]`

- **Preview Label**: If `usesReference` is `["product"]`, it will show preview label when Product Context is available on Frontend but not on Editor side.

Now there are 2 cases:
1. If product context isn't available, we will show following UI in Inspector Control:
![image](https://github.com/user-attachments/assets/aeb4e762-5c22-4e9f-8df9-053793a3c93f)

2. When product context can be inherited from templates (i.e. Single Product Template, Specific Single Product Template) or block (i.e. when collection is inside a Single Product Block). In this case we will show following UI in Inspector Control:
![image](https://github.com/user-attachments/assets/e900df4c-1bb6-4918-91ec-98c90f34daeb)

The first case, where a product is manually selected, has already been addressed in PR [#50590](https://github.com/woocommerce/woocommerce/pull/50590), which has been merged. For the second case, where product context is inherited, @kmanijak has a draft PR #51091 in progress. This PR is expected to be merged soon, which will complete the implementation of both scenarios.

**How above 2 cases will work for "Related Products" collection?**

- When product is manually selected, we will have product ID in [query.productReference](https://github.com/woocommerce/woocommerce/blob/71ed41c2736ea3fefea00609c4bc66d05a5ae06e/plugins/woocommerce-blocks/assets/js/blocks/product-collection/types.ts#L112).
    - On Editor side, we use [update_rest_query_in_editor](https://github.com/woocommerce/woocommerce/blob/58d5c95dfeb2e73fd96dd36044fc7ff172bb4c30/plugins/woocommerce/src/Blocks/BlockTypes/ProductCollection.php#L592) function to update final WP_Query arguments. I can see that you have already made some changes to this in PR. But I believe we will need to use `query.productReference` instead of `woocommerceRelatedTo` to get product ID. IMO Upsells and Cross-sells should also use this approach.
    - On Frontend also we will need to use `query.productReference` to get product ID. I can see that you already made changes in [get_final_frontend_query](https://github.com/woocommerce/woocommerce/blob/58d5c95dfeb2e73fd96dd36044fc7ff172bb4c30/plugins/woocommerce/src/Blocks/BlockTypes/ProductCollection.php#L701). But here also, I believe we will need to use `query.productReference` instead of `woocommerceRelatedTo`.
- When product context is inherited (i.e. "From the current product"), we need to consider two scenarios:
    - When we have a known product ID on Editor side. For example, in case of Specific Single Product Template or when collection is inside a Single Product Block. In both of these cases, we have known product ID.
        - On Editor side, this product ID is passed in [productCollectionLocation](https://github.com/woocommerce/woocommerce/blob/58d5c95dfeb2e73fd96dd36044fc7ff172bb4c30/plugins/woocommerce-blocks/assets/js/blocks/product-template/edit.tsx#L269) while calling REST API. This way, it's accessible in [update_rest_query_in_editor](https://github.com/woocommerce/woocommerce/blob/58d5c95dfeb2e73fd96dd36044fc7ff172bb4c30/plugins/woocommerce/src/Blocks/BlockTypes/ProductCollection.php#L592) function using `$request->get_param( 'productCollectionLocation' )`.
        - On Frontend side as well, product ID is available as part of [productCollectionLocation](https://github.com/woocommerce/woocommerce/blob/58d5c95dfeb2e73fd96dd36044fc7ff172bb4c30/plugins/woocommerce/src/Blocks/BlockTypes/ProductCollection.php#L206).
        - In short, we have product ID available for use in both Editor and Frontend.
    - When we don't have a known product ID on Editor side. For example, inside Single Product Template. product ID will be available on Frontend as part of [productCollectionLocation](https://github.com/woocommerce/woocommerce/blob/58d5c95dfeb2e73fd96dd36044fc7ff172bb4c30/plugins/woocommerce/src/Blocks/BlockTypes/ProductCollection.php#L206) but it's not possible to calculate it on Editor side.
        - In this case, we will show preview label as I mentioned earlier in this comment. Which will work out of the box with `usesReference` to `["product"]`.

But there is one more questions worth asking.

**Do we need to store which option is currently selected by user? i.e. "From the current product" or "From a specific product"**

Another way to ask this is how do we know that "From the current product" option is selected? or "From a specific product" option is selected?

I feel like we will need to have a boolean attribute like `isFromCurrentProduct` or something like that to indicate that "From the current product" option is selected. But this may create problems for Sync patterns. Because we don't want to have contextual attributes for sync patterns. I would love to hear your thoughts on this.

Alternatively, we can just check value of `query.productReference`. If it's available, we can assume that "From a specific product" option is selected. Otherwise, we can assume that "From the current product" option is selected. What do you think? I am inclining towards this approach as this doesn't lead to creation of new attributes and therefore we won't have any issues in Sync patterns.

Maybe @kmanijak can provide more input on this because he is working on [#50916](https://github.com/woocommerce/woocommerce/issues/50916) which adds these two options.

---

I apologize for the lengthy comment, but I wanted to provide a comprehensive overview of my thoughts on how an ideal "Related Products" collection should function. I'd greatly appreciate your feedback and insights, as there may be aspects I've overlooked or misunderstood 🙇‍♂️

ObliviousHarmony committed:
Merge branch 'trunk' into add/49335-related-products-collection

ObliviousHarmony commented:
Thanks for the thorough insight @imanish003!

> I believe we will need to use query.productReference instead of `woocommerceRelatedTo` to get product ID

Yes, this is correct. I tried to make this PR flexible enough that `woocommerceRelatedTo` could be easily replaced or populated in some dynamic way without reworking it. Even the tests will work as-is :smile:

> Do we need to store which option is currently selected by user? i.e. "From the current product" or "From a specific product"

This same type of question came up when I was considering the configuration options for Related Products. The `query` attribute is intended to be passed to `WP_Query` and contains the query paramters to execute. While this makes a ton of sense for the filters and simple collections, when it comes to more complicated ones, I'm not sure that we should be storing block configuration state in the `query` attribute. _Normally_ this wouldn't matter since you can just add block attributes but we're using the variations API and so collections can't create new attributes.

I was playing around with the idea of some sort of unstructured `collectionConfig` object attribute. This would contain any state that the custom collection needs to serialize. Maybe `customConfig` would work better? This would be a good place to store state that is a part of the block's interface or function but not a part of the query.

> Because we don't want to have contextual attributes for sync patterns.

Yes, however, I don't think this is a contextual attribute so much as a configuration option for the block's inspector.

> Alternatively, we can just check value of query.productReference. If it's available, we can assume that "From a specific product" option is selected. Otherwise, we can assume that "From the current product" option is selected. What do you think? I am inclining towards this approach as this doesn't lead to creation of new attributes and therefore we won't have any issues in Sync patterns.

I'm concerned that an empty `query.productReference` has a specific meaning already. Having set `usesReference: [ 'product' ]` means that I should expecft that attribute to exist and contain whatever selected I have made for the product. If it's empty, I would expect that it means I haven't selected a product yet.

kmanijak commented on plugins/woocommerce-blocks/assets/js/blocks/product-collection/collections/related-to.tsx:
```diff
@@ -0,0 +1,63 @@
+	title: __( 'Related Products', 'woocommerce' ),
+	icon: <Icon icon={ loop } />,
+	description: __( 'Recommend products like this one.', 'woocommerce' ),
+	keywords: [ 'related', 'product collection' ],
```
Nitpick: words used in a title are already a keywords so we can skip `related`

```suggestion
	keywords: [ 'product collection' ],
```

ObliviousHarmony replied:
Fixed in 7aeb4052d208ffcd3ae810a466692087600e4a7f

kmanijak reviewed (commented):
I'm curious about the intended way of using this collection. There's `query.woocommerceRelatedTo` string array added to query but I'm not sure I get how it could be used.

My impression is this collection should work only based on:
- collection name - `woocommerce/product-collection/related` would determine it should fetch related products
- `usesReference` - should provide a reference for a product (manually chosen or "current product" soon).

The only query attribute I anticipate to be added in next iteration is "related by" coming from this control but I assume it's not in the scope yet:

<img width="259" alt="image" src="https://github.com/user-attachments/assets/2cf084f6-f1a1-4534-bad7-821d95718244">

What do I miss 😅 What would be the example values of `query.woocommerceRelatedTo`?

---

EDIT: I missed @imanish003's comment above but I think we're touching on similar aspects 🙌

imanish003 commented:
Hey @ObliviousHarmony,

> I tried to make this PR flexible enough that woocommerceRelatedTo could be easily replaced or populated in some dynamic way without reworking it. Even the tests will work as-is 😄
> 

Do we agree that we can use `productReference` to get product ID in both Editor and Frontend for now? Later, if needed, we can add a new attribute. What do you think? 🙂

**Sidenote:** IMO Even if we will need a new attribute, I don't think it should be added at Product Collection block level. Instead, we will need to add it at collection level. Because it's for this specific collection. What do you think?

> This same type of question came up when I was considering the configuration options for Related Products. The query attribute is intended to be passed to WP_Query and contains the query paramters to execute. While this makes a ton of sense for the filters and simple collections, when it comes to more complicated ones, I'm not sure that we should be storing block configuration state in the query attribute.
Normally this wouldn't matter since you can just add block attributes but we're using the variations API and so collections can't create new attributes.
> 

After reading this message, I looked at how @xristos3490 is adding new attribute for "Frequently Bought Together" collection. Looks like he is also using "query" attribute to store selected value from Inspector Control. (see [[#51076](https://github.com/woocommerce/woocommerce/pull/51076)](https://github.com/woocommerce/woocommerce/pull/51076)). I agree that it makes sense to use `query` attribute for simple controls and collections.

Also, Looks like we also used `query` attribute for our Products beta block, which was built on top of Query Loop block. (see [[this code](https://github.com/woocommerce/woocommerce/blob/58d5c95dfeb2e73fd96dd36044fc7ff172bb4c30/plugins/woocommerce-blocks/assets/js/blocks/product-query/types.ts#L11-L59)](https://github.com/woocommerce/woocommerce/blob/58d5c95dfeb2e73fd96dd36044fc7ff172bb4c30/plugins/woocommerce-blocks/assets/js/blocks/product-query/types.ts#L11-L59))

Regarding, creating a new attribute at collection level. I believe we will need to provide an API for this. I believe one way to achieve this could be using `blocks.registerBlockType` filter. This filter allow us to dynamically add new attributes to the block. Therefore, we can add one attribute per collection using this API. By creating a new API on top of `blocks.registerBlockType` filter, we can make it more easy and intuitive for developers to use it. We are already using this API [[here](https://github.com/woocommerce/woocommerce/blob/58d5c95dfeb2e73fd96dd36044fc7ff172bb4c30/plugins/woocommerce-blocks/assets/js/blocks/product-collection/variations/index.ts#L16-L41)](https://github.com/woocommerce/woocommerce/blob/58d5c95dfeb2e73fd96dd36044fc7ff172bb4c30/plugins/woocommerce-blocks/assets/js/blocks/product-collection/variations/index.ts#L16-L41) to add new `__woocommerceNamespace` attribute to core blocks like `core/post-title` to create "Product Title" variation.

> I was playing around with the idea of some sort of unstructured collectionConfig object attribute. This would contain any state that the custom collection needs to serialize. Maybe customConfig would work better? This would be a good place to store state that is a part of the block's interface or function but not a part of the query.
> 

I like the idea. To build on top of what I shared earlier, there are two options:

- As you mentioned, we can create a new attribute called `collectionConfig`, which can be used to store custom data for collections. Probably, for each collection, we can have a `key` for each collection name. This way, we can create separation of concern.
- OR, we can add new attribute for each collection. Each attribute name could be generated using collection name.

From both of these, I like your approach, as this doesn't lead to creating too many new attributes 🚀 But let me know your thoughts on this 🙇‍♂️

**Sidenote:** I don't think this is super important for this PR. Unless, we really need to store some data for this collection in an attribute? 🤔

> Yes, however, I don't think this is a contextual attribute so much as a configuration option for the block's inspector.
> 

There are two options

- From the current product
- From a specific product

Let's say we have an attribute named `isFromCurrentProduct` with value `true`. This means that "From the current product" option is selected.

Now if we create sync pattern and add this sync pattern to a new post. This post will also have `isFromCurrentProduct` attribute with value `true`. Which means "From the current product" option is selected. But this isn't true as it's not possible to have "From the current product" in a post. Therefore, I think it's contextual attribute.

Probably, we can implement a fallback mechanism i.e.

- First check if it's possible to have `From the current product` option selected. If not, we should use `From a specific product` option.
    - But this doesn't change the value of `isFromCurrentProduct` attribute. Therefore, if 3PDs are using this attribute, they will face issues. We can't trust the value of this attribute.

Therefore, I feel like we shouldn't expose it as an attribute unless it's specific to a collection. But it's not specific to a collection.

CC: @kmanijak What do you think? As you will be working on [[this task](https://github.com/woocommerce/woocommerce/issues/50916)](https://github.com/woocommerce/woocommerce/issues/50916), this might be a good time to discuss it 😅

> I'm concerned that an empty query.productReference has a specific meaning already. Having set usesReference: [ 'product' ] means that I should expecft that attribute to exist and contain whatever selected I have made for the product. If it's empty, I would expect that it means I haven't selected a product yet.
> 

I am 100% with you on this. But what does it mean that "I haven't selected a product yet"? If a collection is using `usesReference: [ 'product' ]`, it means that it will rely on product to work properly. If `query.productReference` is empty, it will need to look somewhere else to get the product ID. Therefore, collection can check if this ID exist in `productCollectionLocation`. `productCollectionLocation` will have product ID only if it's provided by either Single Product Template or Single Product Block.

Does it make sense? Probably, we need to improve DX here 😅 But I think, this isn't in scope of this PR. If needed, we can create a new issue to improve DX for this case 🙂 What do you think?

CC: @xristos3490 Your input might be valuable here 🙇‍♂️

xristos3490 reviewed (dismissed):
I agree with @kmanijak and @imanish003 here regarding the new query attribute. 

In my view, we don't need the extra `query.woocommerceRelatedTo` key here, since:
- The variation name should indicate the purpose of the query (find related products), and;
- The usesReference API should be handling the context (the currently viewed product.)

Happy to be wrong of course, but I think this is the most obvious path. Thoughts?

xristos3490 commented:
Moreover

> This pull request introduces a new woocommerceRelatedTo query parameter which generates a list of products related to any of the product IDs provided

@ObliviousHarmony I believe your intention here is to establish a deterministic approach for the query attribute, but I think this conflicts with the concept of "contextuality." Specifically, when dealing with the "currently viewed product," we can only store the "flag" attribute to indicate that the data will be filled dynamically. In this scenario, the "flag" refers to the variation name itself. Does that make sense?

ObliviousHarmony committed:
Removed Unnecessary Keyword

ObliviousHarmony committed:
Support Unlimited `wc_get_related_products()`

This changes related product fetching so that `-1` will return all
related products. I also removed a second +10 offset that seems
to have been accidentally added eight years ago.

ObliviousHarmony commented:
Thanks @xristos3490,

I had too many Slack discussions about this PR that didn't end up here 😓

The `woocommerceRelatedTo` attribute only existed so that I could build everything out without dependent PRs needing to be merged.

>  I believe your intention here is to establish a deterministic approach for the query attribute, but I think this conflicts with the concept of "contextuality."

Yes, that's how I was looking at the query. It seems intuitive, especially since it gets serialized, that the query would be deterministic. We toss it at both the client via dynamic rendering and the editor via the REST API. I didn't fully understand the implications of the editor's querying though and so I see now how preview mode is supposed to address the lack of context in the editor.

With all of that in mind, I understand now why using the collection name is ideal. In the interest of readability, I'm working on an a potential implementation of extensibility (internal and without WordPress filters of any kind) to make that very easy and re-usable. I know there's an open discussion about extensibility and I'll comment on that once I have this in place here.

ObliviousHarmony committed:
Removed `woocommerceRelatedTo` Param

Since we're going to use the collection name, we don't need
this functionality any longer.

ObliviousHarmony committed:
Simplified `post__in` Filtering

Since `merge_queries` was already using an intersection merge for
`post__in`, this second function is unnecessary. I've removed it
as well as refactored the merging logic.

ObliviousHarmony committed:
Pass Collection Name To Final Query

ObliviousHarmony committed:
Merge branch 'trunk' into add/49335-related-products-collection

xristos3490 commented on plugins/woocommerce/src/Blocks/BlockTypes/ProductCollection.php:
```diff
@@ -774,10 +780,28 @@ private function get_final_query_args( $common_query_values, $query, $is_exclude
+	 * @return array The collection-specific query to merge.
+	 */
+	private function get_core_collection_query( $collection, $query ) {
+		$collection = preg_match( '/^woocommerce\/product-collection\/(.*)/', $collection, $matches ) ? $matches[1] : '';
```
Loving this approach! It helps illustrate the impact of each collection on the query!
One suggestion: Instead of `preg_match` I'd use `strpos`?

As per PHP [docs](https://www.php.net/manual/en/function.preg-match.php):
```
Do not use preg_match() if you only want to check if one string is contained in another string. Use [strpos()](https://www.php.net/manual/en/function.strpos.php) instead as it will be faster.
````

ObliviousHarmony replied:
I was lazy so I used `preg_match()` to get the collection name without bothering to check the front of the string and remove the first n characters 😅 I pushed up a new approach that I'm _super_ happy with though so this is gone lol

ObliviousHarmony committed:
Added Custom Collection Handlers

Developers can register collections along with handlers that implement
the custom behavior.

ObliviousHarmony committed:
Removed Second Test Query Construction

As it was, `build_query_vars_from_query_block()` was triggering the
ProductCollection instance hooked into WordPress. After that, we
called `build_frontend_query()` on our test instance. This caused some
weird behavior in tests. As a result, however, the tax_query merge
test couldn't rely on the tax_query transformation done by WordPress.

ObliviousHarmony committed:
Added Related Product Collection Handlers

ObliviousHarmony committed:
Revert "Support Unlimited `wc_get_related_products()`"

This reverts commit 41c837202ebad27031b82f4227cba0b307bc7905.

ObliviousHarmony committed:
Fixed Preview Mode Query

It looks like we were checking the wrong place for the preview
state.

ObliviousHarmony committed:
Merge branch 'trunk' into add/49335-related-products-collection

ObliviousHarmony committed:
Fixed Test

ObliviousHarmony committed:
Merge branch 'trunk' into add/49335-related-products-collection

ObliviousHarmony committed:
Updated Collection Heading

kmanijak commented on plugins/woocommerce/src/Blocks/BlockTypes/ProductCollection.php:
```diff
@@ -1681,4 +1787,80 @@ protected function adjust_price_filter_for_tax_class( $price_filter, $tax_class
+	 *
+	 * @throws \InvalidArgumentException If collection handlers are already registered for the given collection name.
+	 */
+	protected function register_collection_handlers( $collection_name, $build_query, $frontend_args = null, $editor_args = null, $preview_query = null ) {
```
❤️

kmanijak reviewed (dismissed):
Nice, amazing work on this @ObliviousHarmony!

I tested the collection and works as expected. Also, considering the number of changes, I smoke tested other collections in couple of scenarios, didn't spot any regression or anything!

The only thing I saw is the issue you've already mentioned, when PC block is placed inside Single Product block and the first item is the same as the reference. But we discussed that elsewhere.

Nice work, for real! 🚀

kmanijak reviewed (commented):
BTW, I initially approved the PR enthusiastically but considering the number of changes in `ProductCollection.php` I'd love someone with bigger backend experience to take a look as well before merging, for example @xristos3490 😅 Sorry for revoking the approval, there's a green light from my side but I believe there should be at least second opinion on that!

ObliviousHarmony committed:
Merge branch 'trunk' into add/49335-related-products-collection

ObliviousHarmony committed:
Better Collection Test Teardown

Passing a callable back means that we can use variables in the
setUp and then the tearDown without any confusion.

ObliviousHarmony committed:
Removed Unnecessary Test Provider

ObliviousHarmony committed:
Fixed `$post__in` Merging

When the intersection is empty it was returning all products. This
makes it so that it returns nothing when there's no results.

imanish003 commented on plugins/woocommerce-blocks/assets/js/blocks/product-collection/collections/related.tsx:
I noticed that configuration for new "Related Products" collection is different from [existing one](https://github.com/woocommerce/woocommerce/blob/add/49335-related-products-collection/plugins/woocommerce-blocks/assets/js/blocks/product-query/variations/related-products.tsx). I am curious if we should align them? 🤔

For example:
| Configuration | Older Version | New Version |
|---------------|---------------|-------------|
| `perPage`     | 5             | 4           |
| `pages`       | 0             | 1           |

Maybe we can request some feedback from Jill around it? What do you think? I don't have a strong opinion on this, but I believe it's worth discussing it 🤔

CC: @kmanijak

ObliviousHarmony replied:
Ultimately these are fully customizable things, right? We're just showing them an initial state of the block and if they want different pagination customization then they can very easily change that. I think in this case it's actually fine to deviate as long as 4 products _looks better_.

imanish003 replied:
Right! FWIW Here is how New vs Old version of "Related Products" block looks like in Editor:
New Version: We are using "responsive" as "true," so it doesn't enforce a number of columns, which is actually looking good.
![image](https://github.com/user-attachments/assets/242c7d0b-67fc-49e2-a3b3-505cb4f09ae7)


Also, we will have another opportunity to make changes to this configuration when we will soft-deprecate old "Related Products" with this new one. Therefore, I think it's okay to merge this PR as-is.

CC: @j111q

imanish003 commented on plugins/woocommerce/src/Blocks/BlockTypes/ProductCollection.php:
```diff
@@ -18,6 +19,13 @@ class ProductCollection extends AbstractBlock {
+	 * An array keyed by the name of the collection containing handlers for implementing custom collection behavior.
+	 *
+	 * @var array
+	 */
```
Nit: Just a random suggestion while going through the code. Feel free to ignore 😅

```suggestion
		/**
	 * An associative array of collection handlers.
	 *
	 * @var array<string, callable> $collection_handler_store
	 * Keys are collection names, values are callable handlers for custom collection behavior.
	 */
```

imanish003 commented on plugins/woocommerce/src/Blocks/BlockTypes/ProductCollection.php:
```diff
@@ -123,6 +131,8 @@ function ( $html ) {
 		// Disable client-side-navigation if incompatible blocks are detected.
 		add_filter( 'render_block_data', array( $this, 'disable_enhanced_pagination' ), 10, 1 );
+
+		$this->register_core_collections();
```
One concern I have is that we're not utilizing the same extensibility points for our core collections that are currently used by 3PDs. However, this could actually be an opportunity. By experimenting with our core collections, we can potentially discover and refine better extensibility points that could later benefit 3PDs 🤔

ObliviousHarmony replied:
I'm going to mark this as resolved since it's a duplicate of a comment in the top-level response. I'll respond to it there for organization :)

xristos3490 commented on plugins/woocommerce/src/Blocks/BlockTypes/ProductCollection.php:
```diff
@@ -1688,4 +1799,80 @@ protected function adjust_price_filter_for_tax_class( $price_filter, $tax_class
+	 *
+	 * @throws \InvalidArgumentException If collection handlers are already registered for the given collection name.
+	 */
+	protected function register_collection_handlers( $collection_name, $build_query, $frontend_args = null, $editor_args = null, $preview_query = null ) {
```
@ObliviousHarmony Given the environment we are currently working in (a single huge PHP class), this abstraction is great! 

However, I feel that we are assuming that the editor preview will always have the exact same needs as the frontend. While this is true in most cases, I can't help but see that this is a bit more complex than necessary. This is because the flow introduces one more logical leap: some callbacks will be fed into arguments into others, effectively introducing another layer.

What if we only have three arguments here? (a) build_frontend_query, (b) build_editor_query, (c - optional) build the preview query?

With that, I can already see @kmanijak's's [pull request on upsells](https://github.com/woocommerce/woocommerce/pull/51374/files#diff-933825fdae3edb1301562e70acef6627856dd1c39c7475056ef21ac4f0ee9941R1917) repeating the same logic to grab the "collection arguments" while we could introduce a few helpers in the ProductCollectionUtils class.

For example, instead of:
```php
$this->register_collection_handlers(
    build_query_fn,
    get_frontend_args,
    get_editor_args,
    get_preview_args,
);
```
Which will end up repeating much of the code for every collection that uses a "viewed" entity to:
```php
$this->register_collection_handlers(
    build_frontend_query,
    build_editor_query,
    build_preview_query (optional)
);
```

And at the same time within the `ProductCollectionUtils` class:
```php
public function get_viewed_product( $block ) {
	$location = $block->context['productCollectionLocation'] ?? false;
	if ( ! $location || ! is_array( $location ) || ! isset( $location['type'] ) || 'product' !== $location['type'] || ! isset( $location['sourceData'] ) || empty( $location['sourceData']['productId'] ) ) {
		return false;
	}

	$product = wc_get_product( $location['sourceData']['productId'] );
	if ( ! ( $product instanceof WC_Product ) ) {
		return false;
	}

	return $product;
}
```

and a similar for the editor side, like:

```php
public function rest_get_viewed_product( WP_REST_Request ) {
    //...
    return false || WC_Product;
}
```

That way this case would be:
```php
$this->register_collection_handlers(
    function( $query, $block ) {
        $product = ProductCollectionUtils::get_viewed_product( $block );
        if ( ! $product ) {
            return $query; // or prevent results.
        }
        
        return array(
            'post__in' => wc_get_related_products(
            $product->get_id(),
            100
       ),
    function( $query, $request ) {
        $product = ProductCollectionUtils::rest_get_viewed_product( $request );
        if ( ! $product ) {
            return $query; // or prevent results.
        }
        
        return array(
            'post__in' => wc_get_related_products(
            $product->get_id(),
            100
       ),
    );
}
```

What do you think? These helpers can later be used by 3PDs as well, I guess, or be a concrete example of how to use "viewed" entities in a standard way. 

I don't have a strong preference, just sharing my vision here. Feel free to question my thinking :)

imanish003 replied:
I was also reviewing "Upsells" collection and I also noticed that we are repeating same logic. I don't have strong opinion for signature of `register_collection_handlers` but I think it would great to have utility functions in `ProductCollectionUtils` class. Although, I don't think this needs to be in scope of this PR. We can probably create a new issue for this and merge this PR 🙂

ObliviousHarmony replied:
While this is a great discussion @xristos3490, I think we should have this in the discussion thread related to extensibility. At least in our case now, it's okay because this is all internal and subject to change.

ObliviousHarmony replied:
As for moving it to `ProductCollectionUtils`, I think we should have a more specific issue around refactoring the `ProductCollection` class. Dumping everything in a utils class isn't really any better and I think there's a lot of interesting ways we could refactor the collection class. At least for now, I'd like to leave this here.

imanish003 replied:
> While this is a great discussion @xristos3490, I think we should have this in the discussion thread related to extensibility. At least in our case now, it's okay because this is all internal and subject to change.

Agreed! I really like the suggestion from @xristos3490. But It makes sense to not block this PR from merging. We can continue the discussion in the [GH discussion](https://github.com/woocommerce/woocommerce/discussions/51018) and refactor the code if needed once we have some agreement on the extensibility points 🙂

xristos3490 replied:
Thank you, everyone! We don't need to solve everything here; I agree! This class definitely needs a complete refactor. I've added [these thoughts](https://github.com/woocommerce/woocommerce/discussions/51018#discussioncomment-10680769) to the discussion.

imanish003 replied:
> As for moving it to ProductCollectionUtils, I think we should have a more specific issue around refactoring the ProductCollection class. Dumping everything in a utils class isn't really any better and I think there's a lot of interesting ways we could refactor the collection class. At least for now, I'd like to leave this here.

Makes sense to me. I have created a new issue to refactor `ProductCollection` class -- #51534

imanish003 reviewed (changes requested):
I really like the approach you've taken here by creating the `register_collection_handlers` function to register collection handlers! It's a clean, modular way to manage different collection types, which not only improves code readability but also sets us up nicely for future extensibility. This structure could make it easier for both us and third-party developers to add or modify collections down the line. Great job on implementing this 🚀 🙌🏻

I have some minor suggestions and concerns which I will share below. I think we can merge this PR once we have some agreement on these points 🙂

- One concern I have is that we're not utilizing the same extensibility points for our core collections that are currently available for 3PDs. However, this could actually be an opportunity. By experimenting with our core collections, we can potentially discover and refine better extensibility points that could later benefit 3PDs 🤔
	- One thing I liked is that even if we expose new extensibility points, we aren't removing existing ones. It's possible to still use `rest_product_query` and `query_loop_block_query_vars` hooks if needed. But we are providing more granular control over what should happen for each collection type. 
- Also, This PR is changing good amount of code in `ProductCollection` class. Should we consider moving most of collection registration logic to a separate class? OR We can move it to [ProductCollectionUtils.php](https://github.com/woocommerce/woocommerce/blob/add/49335-related-products-collection/plugins/woocommerce/src/Blocks/Utils/ProductCollectionUtils.php). What do you think?
	- I feel like this refactoring is important in case we won't go with this approach for extensibility. I feel refactoring now will save us more time in the long run. What do you think? 🙂

### Issue: When selected product doesn't exist

I tested following scenario:
- Add "Related Products" collection
- Select a product reference
- Using Code Editor, change the product ID to a non-existent product ID
- Save and publish the changes
- Go to the Frontend and notice that Related Products is showing random products.
	- IMO In this case, we should render nothing on Frontend. What do you think? 🤔 I don't think it's a big deal because it's not throwing any errors on Frontend. But still wanted to discuss if we should make any changes.

ObliviousHarmony committed:
Fixed PHPDoc

Co-authored-by: Manish Menaria <the.manish.menaria@gmail.com>

ObliviousHarmony committed:
Hide Empty Related Product Collections

ObliviousHarmony commented:
Thanks @imanish003,

> One concern I have is that we're not utilizing the same extensibility points for our core collections that are currently available for 3PDs. However, this could actually be an opportunity. By experimenting with our core collections, we can potentially discover and refine better extensibility points that could later benefit 3PDs

That's exactly what I'm trying to do :smile:

> I feel like this refactoring is important in case we won't go with this approach for extensibility. I feel refactoring now will save us more time in the long run. What do you think?

I agree with you, however, I think we have a larger problem to solve than the few lines of code this introduces. I think we should create a new issue to look at the `ProductCollection` class and propose a refactor design that we can spend some time working towards. That's not to say I should kick the can down the road, we should prioriotize this work, but I think we should do it in a way that is more thoughtful and deliberate.

> Issue: When selected product doesn't exist

I've fixed this in https://github.com/woocommerce/woocommerce/pull/51076/commits/376a1b5223899c808702f424e156e03d7b6a9a16.

ObliviousHarmony committed:
Merge branch 'trunk' into add/49335-related-products-collection

imanish003 commented:
@ObliviousHarmony Makes sense to me. I have created a new issue to refactor `ProductCollection` class -- #51534

imanish003 committed:
Fix lint issue

imanish003 reviewed (changes requested):
I tested this PR again today and compared the old Related Products block with this new Related Products collection. I noticed that the products weren't the same between them. Therefore, I looked at the old Related Products block [logic](https://github.com/woocommerce/woocommerce/blob/ed9282ffda89c6cb3509b466660ee830d8879a68/plugins/woocommerce/src/Blocks/BlockTypes/RelatedProducts.php#L151-L174) and here is what is different between old vs new one:

- Old one applies visibility filtering using `wc_products_array_filter_visible`
    - IMO this should be applied for new "Related Products" collection as well. What do you think?
- Old one is randomizing products using following code: `wc_products_array_orderby( $related_products, 'rand', 'desc' )`
    - I don't think it makes sense to always randomize products as we allow users to selected "Order By" from Inspector controls. A good solution could be to introduce new "Order by" option, i.e., "Random," to allow users to randomize products. We can select this option by default. There is already [existing issue](https://github.com/woocommerce/woocommerce/issues/49530) for this.
- The old one excludes upsell products, probably because it was always used with the "Upsells" block. But this won't be case with new "Related Products" collection.
    - IMO, it doesn't always make sense to exclude upsell products. But probably, we will need to provide a control for this in the future. Because when "Related Products" will be used with "Upsells" block, merchant will want to exclude already shown upsell products.

In conclusion, I don't want to push any changes to this PR without discussing above points first. I will wait for your thoughts on this 🙂

imanish003 commented:
As discussed on Slack(p1726830742957769-slack-C05UMTA7W0P), I am merging this PR. We can create follow-up PRs to make changes as needed 🙌🏻
